[release/2.1] Add `@(ProjectReference)` items for patched projects

dougbu opened the pull request:
- #29023
- ensure tests do not use baseline assemblies for projects in current patch

----

Originally was an attempt to demonstrate #29023
- showed test indirect references let most tests keep working

dougbu commented:
@Tratcher I believe I've confirmed #29023 is a real issue but would appreciate your confirmation. With the current state of this PR, I'd expect some Kestrel tests (for example) to fail but all I see listed in the Tests tab are projects with direct Microsoft.AspNetCore.Http references. Do you agree many more tests should have failed❔

Tratcher commented:
Lol, you intentionally broke cookies? I would not expect any kestrel tests to fail, but I'm surprised I don't see any failures in Security.

dougbu commented:
> Lol, you intentionally broke cookies?

😏

dougbu commented:
:ok: the binary log 💯% confirmed Microsoft.AspNetCore.Http.dll was resolved from the baseline package in every project _except_ test projects containing `<Reference Include="Microsoft.AspNetCore.Http" />`. That isn't what we want and I'm starting work on resolving #29023.

/fyi @dotnet/aspnet-build

dougbu committed:
Add `@(ProjectReference)` items for patched projects
- #29023
- ensure tests do not use baseline assemblies for projects in current patch

dougbu commented on eng/targets/ResolveReferences.targets:
```diff
@@ -148,6 +148,68 @@
+      <_ProjectsToAdd Remove="@(_ProjectsToAdd)" />
+      <_ProjectsToAdd Include="@(_ProjectsInPatch)"
+          Condition=" @(RuntimeCopyLocalItems->WithMetadataValue('Filename', '%(_ProjectsToAdd.Identity)')->Count()) != 0 OR
+               @(ResolvedCompileFileDefinitions->WithMetadataValue('Filename', '%(_ProjectsToAdd.Identity)')->Count()) != 0 " />
```
@rainersigwald I'm not able to get these item functions to do what I expect. Suggestions much appreciated❕ Note the target is _not_ batched.

dougbu replied:
Idea is to check through `@(RuntimeCopyLocalItems)` and `@(ResolvedCompileFileDefinitions)` for items with filenames matching item specs in `@(_ProjectsToAdd)`.

I didn't do intersection using `<JoinItems>` or like https://github.com/dotnet/aspnetcore/blob/main/eng/targets/ResolveReferences.targets#L96-L125 because the requirements _should_ be simpler here. We don't need any metadata from either  `@(RuntimeCopyLocalItems)` or `@(ResolvedCompileFileDefinitions)` items.

I could create properties for filenames of the two item groups then put e.g. `$(RuntimeCopyLocalItems.Contains(' %(Identity);'))` in here. But, that seems like an unnecessary indirection. Hoping there's a better approach&hellip;

dougbu commented on eng/targets/ResolveReferences.targets:
```diff
@@ -148,6 +148,68 @@
+    <Message Importance="high" Text="ResolvedCompileFileDefinitions / 'Microsoft.AspNetCore.Http': &apos;@(ResolvedCompileFileDefinitions->WithMetadataValue('Filename', 'Microsoft.AspNetCore.Http'))&apos;"
+        Condition=" '$(MSBuildProjectName)' == 'ApplicationInsightsHostingStartupSample' " />
+    <Message Importance="high" Text="ResolvedCompileFileDefinitions / %(_ProjectsInPatch.Identity): &apos;@(ResolvedCompileFileDefinitions->WithMetadataValue('Filename', '%(_ProjectsInPatch.Identity)'))&apos;"
+        Condition=" '$(MSBuildProjectName)' == 'ApplicationInsightsHostingStartupSample' " />
```
These tasks output the following, indicating the Microsoft.AspNetCore.Http filename exists in both groups but isn't found when using `%(_ProjectsInPatch.Identity)`:
``` text
...
  lection;System.Reflection.Metadata;System.Reflection.Primitives;System.Reflection.TypeExtensions;System.Resources.Res
  ourceManager;System.Runtime;System.Runtime.CompilerServices.Unsafe;System.Runtime.Extensions;System.Runtime.Handles;S
  ystem.Runtime.InteropServices;System.Runtime.InteropServices.RuntimeInformation;System.Runtime.Numerics;System.Securi
  ty.AccessControl;System.Security.Cryptography.Algorithms;System.Security.Cryptography.Cng;System.Security.Cryptograph
  y.Encoding;System.Security.Cryptography.Primitives;System.Security.Cryptography.X509Certificates;System.Security.Cryp
  tography.Xml;System.Security.Permissions;System.Security.Principal.Windows;System.Text.Encoding;System.Text.Encoding.
  Extensions;System.Text.Encodings.Web;System.Threading;System.Threading.Tasks;System.Threading.Tasks.Parallel;System.V
  alueTuple;System.Xml.ReaderWriter;System.Xml.XDocument'
  ResolvedCompileFileDefinitions / 'Microsoft.AspNetCore.Http': 'C:\Users\dougbu\.nuget\packages\microsoft.aspnetcore.h
  ttp\2.1.1\lib\netstandard2.0\Microsoft.AspNetCore.Http.dll'
  ResolvedCompileFileDefinitions / Microsoft.AspNetCore.DataProtection.AzureKeyVault: ''
  ResolvedCompileFileDefinitions / Microsoft.AspNetCore.DataProtection.AzureStorage: ''
  ResolvedCompileFileDefinitions / Microsoft.AspNetCore.Http: ''
```

Output for the `@(RuntimeCopyLocalItems)` item group is similar.

dougbu commented:
With help from @nmytelka and @yuehuang010, this new target is now working. Many thanks❕

I'm going to try out a couple of tweaks locally then back out the first two commits (i.e. don't intentionally break tests) and turn this into a real PR😃

dougbu commented:
Note latest iteration failed many more tests than previous ones i.e. transitive references to the project I broke are no longer hiding the break from test projects and samples.

dougbu commented:
/fyi [test run with the fix](https://dev.azure.com/dnceng/public/_build/results?buildId=968833) failed 404 tests on Windows and 322 elsewhere while [run before fix](https://dev.azure.com/dnceng/public/_build/results?buildId=938869&) failed far fewer -- 206 tests on Windows and 103 elsewhere.

dougbu commented:
This is `tell-mode` because it's infrastructure and the changes affect only **non**-implementation projects.

dougbu commented on eng/targets/ResolveReferences.targets:
```diff
@@ -148,6 +148,60 @@
+
+      <!-- Ignore unreferenced assemblies and directly-referenced projects. -->
+      <_ProjectsToAdd Remove="@(_ProjectsToAdd)" />
+      <!-- Condition uses same Filename metadata when evaluating the other two item groups. -->
```
Put another way, the `Condition`s here and at the end of the target use a corner case of the task batching features. Shared metadata and batching on that metadata are the keys.

wtgodbe commented on eng/targets/ResolveReferences.targets:
```diff
@@ -148,6 +148,60 @@
+      <_ProjectsToAdd Remove="@(_ProjectsToAdd)" />
+      <!-- Condition uses same Filename metadata when evaluating the other two item groups. -->
+      <_ProjectsToAdd Include="@(_ProjectsInPatch)"
+          Condition=" '%(Filename)' != '' AND
```
What is this condition doing exactly? Add a direct reference to the projects being patched, unless...

dougbu replied:
My mental model may be wrong (@Forgind and @yuehuang010, please chime in) but https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-batching?view=vs-2019#item-and-property-mutations contains _some_ of the information
1. The `'%(Filename)` item metadata causes the implicit `CreateItem` task to be batched
2. Use of that metadata also limits the `@(ResolvedCompileFileDefinitions)` and `@(RuntimeCopyLocalItems)` item groups to the item(s) with the same `%(Filename)`
    - This part doesn't seem to be explicit in the documentation
3. The `OR` clause confirms at least one such item exists

At a higher level, this creates a `@(_ProjectsToAdd)` item group containing those `@(_ProjectsInPatch)` items that match on filename with  one or more `@(ResolvedCompileFileDefinitions)` items `@(RuntimeCopyLocalItems)`, excluding existing `@(ProjectReference)` items. Another way of looking at it is the created item group contains the projects in the current patch that (a) aren't referenced directly but (b) match assemblies in the current build.

yuehuang010 replied:
I see it used in this exampled (https://docs.microsoft.com/en-us/visualstudio/msbuild/item-metadata-in-task-batching?view=vs-2019#divide-several-item-lists-into-batches).  So, it is feature name is "Divide several item lists into batches":
``````
    <Target Name="ShowMessage">
        <Message
            Text = "Number: %(Number) -- Items in ExampColl: @(ExampColl) ExampColl2: @(ExampColl2)"/>
    </Target>
``````

dougbu replied:
Good point @yuehuang010. I'm not sure how many times I read that example without thinking "Oh, this makes intersection / subsetting dead easy"😃

dougbu commented:
@wtgodbe is there a reason the aspnetcore-pr-validation-temp pipeline still exists and is still the one we require for release/2.1 merges❔ Tests failed in the first run since I got rid of the intentional break but only in that pipeline. The same tests ran fine in aspnetcore-ci.

wtgodbe commented:
> @wtgodbe is there a reason the aspnetcore-pr-validation-temp pipeline still exists and is still the one we require for release/2.1 merges❔ Tests failed in the first run since I got rid of the intentional break but only in that pipeline. The same tests ran fine in aspnetcore-ci.

No, I can remove the validation-temp runs

wtgodbe commented:
https://github.com/dotnet/aspnetcore/pull/29731
